File: sdk/typescript/src/api/feedback.ts
import type { HttpClient, RequestOptions } from "../http.js";

export type FeedbackType = "feature" | "bug";
export type FeedbackVoteValue = 1 | -1 | 0;

export interface CreateFeedbackBody {
  type: FeedbackType;
  title: string;
  body: string;
}

export interface IngestFeedbackBody {
  type: FeedbackType;
  title: string;
  body: string;
  product: "backend" | "opencompany" | "openhuman" | "medulla";
  origin?: string;
  externalRef?: string;
}

export interface ListFeedbackQuery {
  type?: FeedbackType;
  status?: "open" | "planned" | "completed";
  sort?: "hot" | "top" | "new";
  page?: number;
  limit?: number;
}

export interface CommentFeedbackBody {
  body: string;
}

export interface VoteFeedbackBody {
  value: FeedbackVoteValue;
}

/**
 * Public feedback board: submit and browse feature requests and bug reports,
 * read individual items with comments, vote, and comment.
 */
export class FeedbackApi {
  constructor(private readonly http: HttpClient) {}

  /** Submit feedback or a bug report (LLM-moderated, rate-limited). */
  createFeedback<T = unknown>(body: CreateFeedbackBody, options?: RequestOptions): Promise<T> {
    return this.http.post<T>("/feedback", body, options);
  }

  /** Ingest authenticated product feedback into the moderation pipeline. */
  ingestFeedback<T = unknown>(
    body: IngestFeedbackBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/feedback/ingest", body, options);
  }

  /** List feedback on the public board. */
  listFeedback<T = unknown>(query?: ListFeedbackQuery, options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/feedback", {
      ...options,
      query: { ...query, ...options?.query },
    });
  }

  /** Get a feedback item with its comments. */
  getFeedback<T = unknown>(id: string, options?: RequestOptions): Promise<T> {
    return this.http.get<T>(`/feedback/${encodeURIComponent(id)}`, options);
  }

  /** Comment on a feedback item. */
  commentFeedback<T = unknown>(
    id: string,
    body: CommentFeedbackBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>(
      `/feedback/${encodeURIComponent(id)}/comments`,
      body,
      options,
    );
  }

  /** Up/down-vote a feedback item (value 0 retracts). */
  voteFeedback<T = unknown>(
    id: string,
    body: VoteFeedbackBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>(`/feedback/${encodeURIComponent(id)}/vote`, body, options);
  }
}

File: sdk/python/src/tinyhumans/api/feedback.py
from __future__ import annotations

from typing import Any

from ..http import Json
from ..types import FeedbackIngest
from ._base import ApiNamespace, enc

__all__ = ["FeedbackApi"]


class FeedbackApi(ApiNamespace):
    """Public feedback board: submit, browse, read, vote, and comment."""

    def create_feedback(self, body: dict[str, Any], **kwargs: Any) -> Json:
        """Submit feedback or a bug report (LLM-moderated, rate-limited)."""
        return self._http.post("/feedback", body=body, **kwargs)

    def ingest(self, body: FeedbackIngest, **kwargs: Any) -> Json:
        """Ingest feedback from another TinyHumans product."""
        return self._http.post("/feedback/ingest", body=body, **kwargs)

    def list_feedback(self, query: dict[str, Any] | None = None, **kwargs: Any) -> Json:
        """List feedback on the public board."""
        merged = self._merge_query(query, kwargs.pop("query", None))
        return self._http.get("/feedback", query=merged, **kwargs)

    def get_feedback(self, id: str, **kwargs: Any) -> Json:
        """Get a feedback item with its comments."""
        return self._http.get(f"/feedback/{enc(id)}", **kwargs)

    def comment_feedback(self, id: str, body: dict[str, Any], **kwargs: Any) -> Json:
        """Comment on a feedback item."""
        return self._http.post(f"/feedback/{enc(id)}/comments", body=body, **kwargs)

    def vote_feedback(self, id: str, body: dict[str, Any], **kwargs: Any) -> Json:
        """Up/down-vote a feedback item (value 0 retracts)."""
        return self._http.post(f"/feedback/{enc(id)}/vote", body=body, **kwargs)
